### ISSUE TITLE
Refactor: Seclude "lib/utils/aws/request.js" (AWS request logic) util

### ISSUE THREAD

**Issue by medikoo** [serverless/serverless#8445]
### Use case description

Currently [AWS `request`](https://github.com/serverless/serverless/blob/2d6824cde531ba56758f441b39b5ab018702e866/lib/plugins/aws/provider/awsProvider.js#L1065-L1198) logic is integrated into AWS provider plugin instance, which is part of Serverless instance.

That makes it difficult to rely on it independently, also doesn't allow to easily configure a isolated unit tests

### Proposed solution

_~~Depends on https://github.com/serverless/serverless/issues/8444 which needs to be addressed first~~ (Done!)_

Seclude most of the `request` method (specifically: https://github.com/serverless/serverless/blob/2d6824cde531ba56758f441b39b5ab018702e866/lib/plugins/aws/provider/awsProvider.js#L1070-L1197) into `lib/aws/request.js`), into _non-memoized_ (direct export of `lib/aws/request.js`)  and _memoized_ (exposed on `.memoized` property) with signature `service, method, params = {}, serviceOptions = {}`

Remove dependency on _plugin_ instance:
- Rely on `@serverless/utils/log` instead of `this.serverless.cli.log`
- Integrate logic of `this.canUseS3TransferAcceleration` and `this.enableS3TransferAcceleration` into util and remove them from plugin instance (also remove its related tests)
- Configure `this.requestCache` as `requestCache` module local variable and remove this property from plugin instance
- Configure `this.requestQueue` as `requestQueue` module local variable and remove this property from plugin instance
- Rely on `ServerlessError` as exported by `lib/serverless-error.js` instead of `this.serverless.classes.Error`
- Configure `this.sdk` as `sdk` module local variable and remove this property from plugin instance (refactor only place where [it's used externally](https://github.com/serverless/serverless/blob/2d6824cde531ba56758f441b39b5ab018702e866/lib/plugins/aws/deploy/lib/checkForChanges.js#L193) to rely on `awsProvider.request` method)
- Move [_proxy_ and _timeout_ setup](https://github.com/serverless/serverless/blob/2d6824cde531ba56758f441b39b5ab018702e866/lib/plugins/aws/provider/awsProvider.js#L978-L1034) logic from plugin to body of this module, and remove its related tests from `awsProvider.test.js` (we will also not reintroduce them, as  (1) in given form they were not reliable - were testing against same AWS SDK instance repeatedly (2) it's difficult to configure reliable tests (3) we can just test by following steps of implementation, and not the desired effect)

When writing tests rely on `proxyquire` and `sinon` to configure AWS SDK sinon stub. Proposed tests spec

```javascript
describe('awsProvider.request()`', () => {
  it('should trigger expected AWS SDK invocation', () => {});
  it('should handle subclasses', () => {});
  it('should support region in options set', () => {});
  it('should retry on retryable errors', () => {});
  it('should retry if error code is 429 and retryable is set to false', () => {});
  it('should not retry if error code is 403 and error is retryable', () => {});
  it('should expose non-retryable errors', () => {});
  it('should expose original error message in thrown error message', () => {});
  it('should return ref to docs for missing credentials', () => {});
  it('should enable S3 acceleration if "--aws-s3-accelerate" CLI option is provided', () => {
    // Confirm that passed in service constructor options have "useAccelerateEndpoint" set to true
  });
  it('should support caching through "useCache: true" option', () => {
    // Test that repeated request returns cached result
    // Then test another call with same method but in different region and confirm that no cached value was used
  });
});
```

In `lib/plugins/aws/provider/awsProvider.js` invoke new util internally in `request` method`, and remove all of the `request` method related tests from `awsProvider.test.js`

**Comment by AlinoeDoctari** [comment]
hi @medikoo. It looks like the #8444 was closed, does that mean that it is now possible to get started on that issue? I'm interested for the credential process in ~/.aws/config.

**Comment by medikoo** [comment]
@AlinoeDoctari Absolutely! Great for bringing that up. I simply forgot to update the issue, and mark it with appropriate labels.

Do you want to work on it?

**Comment by AlinoeDoctari** [comment]
I can give it a go yes :)

**Comment by medikoo** [comment]
@AlinoeDoctari thank you! We look forward

**Comment by AlinoeDoctari** [comment]
I've had a look around and tried to move test and functions a bit. I have a first question: part of the extracted code require access to the credentials (at least the S3enableTransferAcceleration and the service creation), is it ok to move the getCredential function into another utils (say utils/credentials) and request it from the utils/request? I think that would make the interactions clearer and I would go down that road, but maybe you had another design in mind @medikoo

**Comment by medikoo** [comment]
@AlinoeDoctari good question. Firstly `credentials` is technically an AWS SDK consturctor options, and all that `canUseS3TransferAcceleration` and `enableS3TransferAcceleration` do, is to set additional option for constructor if request concerns S3 service (so `credentials` input is not really relevant here).

Still to recognize _credentials_ data originally passed with this object, I believe we should accept the options for service constructor in this new request util, and in initial proposal  Ididn't really outlined that.

I've updated it now (check the description).

I've proposed the signature to be: `service, method, params = {}, serviceOptions = {}` and proposed to expose two functions (_memoized_ and _non memoized_), this get rids of original `options` object, which was used to (1) indicate whether we want to result to be cached or not (now consumer simply needs to call different function to have request memoized) (2) customize region (this should really be passed with `serviceOptions`)

So having that sorted that way,  I believe we still need to keep the `request` method (and leave the current usage of it), but it's internals will be simplified to something as:

```javascript
const awsRequest = require('../../../aws/request');

request(service, method, params, options) {
  const serviceOptions = ... // resolve service options
  const shouldCache = ... // resolve should cahce
  return (shouldCache? awsRequest.memoized : awsRequest)(service, method, params, serviceOptions);
}
```
